src/net_processing.cpp

```diff
@@ -584,76 +578,105 @@ bool ProcessMessage(CNode *pfrom, std::string strCommand, CDataStream &vRecv, in
                 pfrom->fDisconnect = true;
             }
         }
-        pfrom->state_incoming = ConnectionStateIncoming::SENT_VERACK_READY_FOR_POTENTIAL_XVERSION;
     }
 
-    /* Since we are processing messages in multiple threads, we may process them out of order.  Does enforcing this
-       order actually matter?  Note we allow mis-order (or no version message at all) if whitelisted...
-        else if (pfrom->nVersion == 0 && !pfrom->fWhitelisted)
+    else if (strCommand == NetMsgType::XVERSION)
+    {
+        if (pfrom->fSuccessfullyConnected == true)
         {
-            // Must have version message before anything else (Although we may send our VERSION before
-            // we receive theirs, it would not be possible to receive their VERACK before their VERSION).
+            dosMan.Misbehaving(pfrom, 1);
             pfrom->fDisconnect = true;
-            return error("%s receieved before VERSION message - disconnecting peer=%s", strCommand,
-       pfrom->GetLogName());
+            return error("odd peer behavior: received verack message before xversion, disconnecting \n");
+        }
+
+        vRecv >> pfrom->xVersion;
+
+        if (pfrom->addrFromPort != 0)
+        {
+            LOG(NET, "Encountered odd node that sent BUVERSION before XVERSION. Ignoring duplicate addrFromPort "
+                     "setting. peer=%s version=%s\n",
+                pfrom->GetLogName(), pfrom->cleanSubVer);
         }
-    */
+
+        pfrom->ReadConfigFromXVersion();
+
+        pfrom->PushMessage(NetMsgType::VERACK);
+    }
+
+    else if ((pfrom->nVersion == 0 || pfrom->tVersionSent < 0) && !pfrom->fWhitelisted)
+    {
+        // Must have a version message before anything else
+        dosMan.Misbehaving(pfrom, 1);
+        pfrom->fDisconnect = true;
+        return error("%s receieved before VERSION message - disconnecting peer=%s", strCommand, pfrom->GetLogName());
+    }
 
     else if (strCommand == NetMsgType::VERACK)
     {
-        if (!ensureConnectionState(
-                strCommand, ConnectionStateIncoming::ANY, ConnectionStateOutgoing::SENT_VERSION, pfrom))
-            return false;
+        if (GetTime() - pfrom->tVersionSent > VERACK_TIMEOUT)
+        {
+            dosMan.Misbehaving(pfrom, 1);
+            return error("verack message receieved after verack timeout limit reached");
+        }
 
```

Comment:
no need to check for verack timeout here...just let it fall through to the next if statement because what's happening is you are no longer disconneting here...this is a potential connection dos vector. If you let it fall through it will get flagged futher downstream and disconnected and the connection logic can update its' counters.